aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -451,6 +451,31 @@ Tensor new_full(
   return at::full(size, fill_value, self.options().merge_in(options));
 }
 
+namespace {
+TensorOptions linspace_logspace_infer_options(
+    Scalar start,
+    Scalar end,
+    const TensorOptions& options) {
+  auto result_options = options;
+  if (start.isComplex() || end.isComplex()) {
```

Comment:
This seems odd. Would you elaborate on what happens in the following cases:

complex start, float end
float start, complex end
complex start, complex end
float start, float end, complex dtype or out
complex start, complex end, float dtype or out
dtype and out tensor type mismatch

Reply:
```python
import torch

def linspace_check(start, end, dtype=None,  device='cpu', out=None):
    if out is None:
        return torch.linspace(start, end, dtype=dtype, device=device)
    if out is not None:
        return torch.linspace(start, end, out=out)

# Start: Imag
# End: Real
# Returns: torch.complex64
t = linspace_check(1j, 2)
print(t.dtype, t.device)
t = linspace_check(1j, 2, device='cuda')
print(t.dtype, t.device)
print('*'*8)

# Start: Real
# End: Imag
# Returns: torch.complex64
t = linspace_check(2, 2j)
print(t.dtype, t.device)
t = linspace_check(2, 2j, device='cuda')
print(t.dtype, t.device)
print('*'*8)

# Start: Imag
# End: Imag
# Returns: torch.complex64
t = linspace_check(1j, 2j)
print(t.dtype, t.device)
t = linspace_check(1j, 2j, device='cuda')
print(t.dtype, t.device)
print('*'*8)

# Start: Real
# End: Real
# Dtype: torch.complex64
# Returns: torch.complex64
t = linspace_check(1, 2, dtype=torch.complex64)
print(t.dtype, t.device)
t = linspace_check(1, 2, dtype=torch.complex64, device='cuda')
print(t.dtype, t.device)
print('*'*8)

# Start: Imag
# End: Imag
# Dtype: torch.float
# Returns: torch.complex64
t = linspace_check(1j, 2j, dtype=torch.float)
print(t.dtype, t.device)
t = linspace_check(1j, 2j, dtype=torch.float, device='cuda')
print(t.dtype, t.device)
print('*'*8)

# Start: Imag
# End: Imag
# out: torch.float
# Returns: Error
# out = torch.zeros(1, dtype=torch.float)
# t = linspace_check(1j, 2j, out=out)
# print(t.dtype, t.device)
# out = out.cuda()
# t = linspace_check(1j, 2j, out=out)
# print(t.dtype, t.device)
# print('*'*8)

# Start: Real
# End: Real
# out: torch.complex64
# Returns: torch.complex64
out = torch.zeros(1, dtype=torch.complex64)
t = linspace_check(1, 2, out=out)
print(t.dtype, t.device)
out = out.cuda()
t = linspace_check(1, 2, out=out)
print(t.dtype, t.device)
print('*'*8)
```

Reply:
Thank you; that's very helpful. Which case(s) throw the warning, imag x imag with out=float?

Reply:
Yup. Warning will be thrown if either `start` or `end` is complex and dtype is not complex.

Reply:
Can we distinguish between dtype=float and a call without dtype set? That is, if start and end are complex and dtype is not set, does it still warn?

Reply:
Unfortunately not. I think codegen makes assumption that if the dtype is not passed for factory functions, it sets them to the default dtype. So the optional dtype this function receives is always set.

Reference: https://github.com/pytorch/pytorch/pull/38875#issuecomment-635481829

Reply:
That's what I thought. I had to implement the custom binding for full in PR https://github.com/pytorch/pytorch/pull/34709 to detect the distinction, and you've correctly identified it above.

Throwing a warning when start and end are complex seems odd. I think we should pursue updating the linspace and logspace bindings to be like full and then add a test that a warning isn't thrown and that a complex start or end errors out when given a float or out dtype.

What do you think, @kshitij12345?

Reply:
I agree. This behaviour is very odd. However would it be ok to pursue it in a follow-up PR?

Reply:
OK, but let's add a test that the warning IS thrown, file an issue, and add a comment linking that issue above the test and in the code.